src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,105 @@
+}
+
+void EngineEffectsDelay::process(const CSAMPLE* M_RESTRICT pIn,
+        CSAMPLE* M_RESTRICT pOutput,
+        const int iBufferSize) {
+    if (m_oldDelaySamples == 0 && m_delaySamples == 0) {
+        SampleUtil::copy(pOutput, pIn, iBufferSize);
+        return;
+    }
+
+    if (m_oldDelaySamples == m_delaySamples) {
+        int delaySourcePos = (m_delayPos + kiMaxDelay - m_delaySamples) % kiMaxDelay;
+
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= static_cast<int>(kiMaxDelay)) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+
+        for (int i = 0; i < iBufferSize; ++i) {
+            // Put samples into delay buffer.
+            m_pDelayBuffer[m_delayPos] = pIn[i];
+            m_delayPos = (m_delayPos + 1) % kiMaxDelay;
+
+            // Take a delayed sample from the delay buffer
+            // and copy it to the destination buffer.
+            pOutput[i] = m_pDelayBuffer[delaySourcePos];
+            delaySourcePos = (delaySourcePos + 1) % kiMaxDelay;
+        }
+
+    } else {
+        VERIFY_OR_DEBUG_ASSERT(m_delaySamples >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+
+        int delaySourcePos = (m_delayPos + kiMaxDelay - m_delaySamples +
+                                     iBufferSize / 2) %
+                kiMaxDelay;
+        int oldDelaySourcePos =
+                (m_delayPos + kiMaxDelay - m_oldDelaySamples) %
+                kiMaxDelay;
+
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= static_cast<int>(kiMaxDelay)) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(oldDelaySourcePos >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(oldDelaySourcePos <= static_cast<int>(kiMaxDelay)) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+
+        double crossMix = 0.0;
+        double crossInc = 2 / static_cast<double>(iBufferSize);
+
+        for (int i = 0; i < iBufferSize; ++i) {
+            // Put samples into delay buffer.
+            m_pDelayBuffer[m_delayPos] = pIn[i];
+            m_delayPos = (m_delayPos + 1) % kiMaxDelay;
+
+            // Take delayed sample from delay buffer and copy it to dest buffer.
+            if (i < iBufferSize / 2) {
+                // Ramp only the second half of the buffer.
+                pOutput[i] = m_pDelayBuffer[oldDelaySourcePos];
+            } else {
+                pOutput[i] = static_cast<CSAMPLE>(
+                        m_pDelayBuffer[oldDelaySourcePos] * (1.0 - crossMix));
```

Comment:
Does this need to be with double precision? I can Imagine that a float calculation is sufficient, but faster on a 32 bit CPU.